src/data/set/prod.lean

```diff
@@ -332,6 +333,59 @@ lemma diag_preimage_prod_self (s : set α) : (λ x, (x, x)) ⁻¹' (s ×ˢ s) =
 
 end diagonal
 
+section off_diag
+variables {α : Type*} {s t : set α} {x : α × α} {a : α}
+
+/-- The off-diagonal of a set `s` is the set of pairs `(a, b)` with `a, b ∈ s` and `a ≠ b`. -/
+def off_diag (s : set α) : set (α × α) := {x | x.1 ∈ s ∧ x.2 ∈ s ∧ x.1 ≠ x.2}
```

Comment:
It's occured to me that this is different from the definition for finset, in a way that actually is quite annoying to fix in my context.  Would you be willing to change this to `{x ∈ s ×ˢ s | x.fst ≠ x.snd}`?

Reply:
(Obviously you have this as a theorem, `off_diag_eq_sep_prod`, but this is (analogous to) the actual definition for finset.

Reply:
There's precedent to always favor right associated `and`s. In that particular case, I can't see how consistency with `finset` matters because you can't use any defeq on the `finset` version.

Reply:
I think this is a fair concern - consistency is important for (many) reasons other than defeq. But I think in this case @linesthatinterlace the `mem_off_diag` lemmas (in both cases) together should make this fairly easy to fix?

Reply:
Yes, I think so... but nevertheless I don't see why this shouldn't be consistent.

Reply:
It's not clear what "consistent" even means here given that you have no defeq on `finset.filter`. To me it seems you have `finset.mem_filter` in mind, or maybe `finset.coe_filter`, but you should never use those on `finset.off_diag` outside of API writing and the API has already been written. Hence I am declaring that the `off_diag` API makes this consistent.

Reply:
I meant that conceptually finset constructs the off_diag by filtering on the product of the two finsets, whereas this constructs it by filtering on the product of the types rather than a sep on the product of sets. But I don't want to hold this up anymore.